### PR TITLE
W6HW10 Optimization of ML models

### PR DESCRIPTION
**Pull request by yevhen-k**
Hi @truskovskiyk 

Here is a huge PR with a lot of code. Sorry, I haven't manage to separate the code into different PRs because the code is logically coupled.

Please, feel free to ask any questions.

### CODE REVIEW

File: week-6-benchmarking/README.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+Results:
+
+```
+---------------- benchmark: 2 tests ----------------
+Name (time in ms)                     Mean          
+----------------------------------------------------
+test_benchmark_forwardpass_gpu      6.8554 (1.0)    
+test_benchmark_forwardpass_cpu     35.3109 (5.15)
```
Comment (truskovskiyk) [inline week-6-benchmarking/README.md:18]: let's add environment here,  benchmarks are not complete without env you used to run them

File: week-6-benchmarking/README.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+make benchmark-api-local
+```
+
+Results:
+
+
+| Type | Name       | # reqs | # fails   | Avg  | Min | Max   | Med  | req/s  | failures/s |
+| ---- | ---------- | ------ | --------- | ---- | --- | ----- | ---- | ------ | ---------- |
+| POST | /          | 2200   | 80(3.64%) | 2988 | 424 | 14541 | 2000 | 108.40 | 1.20       |
+|      | Aggregated | 2200   | 80(3.64%) | 2988 | 424 | 14541 | 2000 | 108.40 | 1.20       |
```
Comment (truskovskiyk) [inline week-6-benchmarking/README.md:60]: same, what was the env: machine type, CPU, RAM, noisy neighbours, etc

File: week-6-benchmarking/api/app.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+
+print(f">>> using device: {device}")
+
+model = RegressionModel()
+model.load_state_dict(torch.load("./weights/bert-base-uncased-regression-weights.pth"))
+model.to(device)
+tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+model.eval()
```
Comment (truskovskiyk) [inline week-6-benchmarking/api/app.py]: would be nice to wrap this block into a function

File: week-6-benchmarking/api/app.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+app = Flask(__name__)
+
+
+@app.route('/', methods=['GET', 'POST'])
```
Comment (truskovskiyk) [inline week-6-benchmarking/api/app.py]: nit: let's add some name to this endpoint

File: week-6-benchmarking/api/app.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        input_ids = tokens["input_ids"].to(device)
+        token_type_ids = tokens["token_type_ids"].to(device)
+        attention_mask = tokens["attention_mask"].to(device)
+        tokens = {
+            "input_ids": input_ids,
+            "token_type_ids": token_type_ids,
+            "attention_mask": attention_mask
+        }
+        with torch.no_grad():
+            response = model(tokens).cpu().numpy().tolist()
```
Comment (truskovskiyk) [inline week-6-benchmarking/api/app.py]: might be a separate method, with  

```
@torch.no_grad()
def inference(....):
    return response

def process():
    .....
    inference()
    .....
```

File: week-6-benchmarking/bert-quntization.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    # create a quantized model instance
+    model.to(device_cpu)
+    model_int8 = torch.quantization.quantize_dynamic(
+        model,  # the original model
+        {torch.nn.LSTM, torch.nn.Linear},  # a set of layers to dynamically quantize
+        dtype=torch.qint8)  # the target dtype for quantized weights
+
+    # check the model size
+    print_size_of_model(model)
+    print_size_of_model(model_int8)
```
Comment (truskovskiyk) [inline week-6-benchmarking/bert-quntization.py:71]: amazing!

**Review by truskovskiyk — COMMENTED** [review]
@yevhen-k amazing work, let's re-base on top of merged PRs - so it would be easier to review

File: week-6-benchmarking/README.md (hunk 1 above)
Comment (yevhen-k) [reply week-6-benchmarking/README.md:18]: Readme updated. In short:

Test set up:
- CPU: Intel Core i5-8300H CPU @ 2.30GHz x 8
- RAM: 15.5GiB
- GPU: NVIDIA GeForce GTX 1050 Ti
- DRAM: 4096 MiB
- CUDA: 11.7
- Noisy neighbours: default minikube cluster activity

File: week-6-benchmarking/README.md (hunk 2 above)
Comment (yevhen-k) [reply week-6-benchmarking/README.md:60]: Readme updated

File: week-6-benchmarking/api/app.py (hunk 3 above)
Comment (yevhen-k) [reply week-6-benchmarking/api/app.py]: done

File: week-6-benchmarking/api/app.py (hunk 4 above)
Comment (yevhen-k) [reply week-6-benchmarking/api/app.py]: done

File: week-6-benchmarking/api/app.py (hunk 5 above)
Comment (yevhen-k) [reply week-6-benchmarking/api/app.py]: done

File: week-6-benchmarking/bert-quntization.py (hunk 6 above)
Comment (yevhen-k) [reply week-6-benchmarking/bert-quntization.py:71]: Thanks!

File: week-6-benchmarking/README.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+test_benchmark_forwardpass_cpu     35.3109 (5.15)   
+----------------------------------------------------
+```
+
+Test set up:
+- CPU: Intel Core i5-8300H CPU @ 2.30GHz x 8
+- RAM: 15.5GiB
+- GPU: NVIDIA GeForce GTX 1050 Ti
+- DRAM: 4096 MiB
+- CUDA: 11.7
```
Comment (truskovskiyk) [inline week-6-benchmarking/README.md:27]: great!

**Review by truskovskiyk — APPROVED** [review]
@yevhen-k great job!

**Pull request merged**